ginga/canvas/types/plots.py

```diff
@@ -0,0 +1,842 @@
+        if y_acc is None:
+            y_acc = np.mean
+        self.y_func = lambda arr: nul_arr if arr.size == 0 else y_acc(arr)
+
+        self.points = np.copy(nul_arr)
+        self.limits = np.array([(0.0, 0.0), (0.0, 0.0)])
+        self.plot_xlim = (None, None)
+
+        self.path = Path([], color=color, linewidth=linewidth,
+                         linestyle=linestyle, alpha=alpha, coord='data')
+        self.path.get_cpoints = self.get_cpoints
+
+    def plot_xy(self, xpts, ypts):
+        """Convenience function for plotting X and Y points that are in
+        separate arrays.
+        """
+        self.plot(np.asarray((xpts, ypts)).T)
+
+    def plot(self, points, limits=None):
+        """Plot `points`, a list, tuple or array of (x, y) points.
+
+        Parameter
+        ---------
+        points : array-like
+            list, tuple or array of (x, y) points
+
+        limits : array-like, optional
+            array of (xmin, ymin), (xmax, ymax)
+
+        Limits will be calculated if not passed in.
+        """
+        self.points = np.asarray(points)
+        self.plot_xlim = (None, None)
+
+        # path starts out by default with the full set of data points
+        # corresponding to the plotted X/Y data
+        #self.path.points = self.get_data_points(points=self.points)
+
+        # set or calculate limits
+        if limits is not None:
+            # passing limits saves costly min/max calculation
+            self.limits = np.asarray(limits)
+        else:
+            self._calc_limits(self.points)
+
+    def _calc_limits(self, points):
+        """Internal routine to calculate the limits of `points`.
+        """
+        # TODO: what should limits be if there are no points?
+        if len(points) == 0:
+            self.limits = np.array([[0.0, 0.0], [0.0, 0.0]])
+        else:
+            x_vals, y_vals = points.T
+            self.limits = np.array([(x_vals.min(), y_vals.min()),
+                                    (x_vals.max(), y_vals.max())])
+
+    def calc_points(self, viewer, start_x, stop_x):
+        """Called when recalculating our path's points.
+        """
+        # in case X axis is flipped
+        start_x, stop_x = min(start_x, stop_x), max(start_x, stop_x)
+
+        new_xlim = (start_x, stop_x)
+        if new_xlim == self.plot_xlim:
+            # X limits are the same, no need to recalculate points
+            return
+
+        self.plot_xlim = new_xlim
+
+        points = self.get_data_points(points=self.points)
+        if len(points) == 0:
+            self.path.points = points
+            return
+
+        x_data, y_data = points.T
+
+        # if we can determine the visible region shown on the plot
+        # limit the points to those within the region
+        if np.all(np.isfinite([start_x, stop_x])):
+            idx = np.logical_and(x_data >= start_x, x_data <= stop_x)
+            points = points[idx]
+
+        if have_npi and self.x_func is not None:
```

Comment:
That's right, I just need `group_by`.  I'll look into `pandas` support for something like that, or something using native Python.

Reply:
@pllim, thanks for pushing back on this one! I think this will do nicely and seems to be just as fast:
```python
            u, i = np.unique(cx, return_index=True)
            gr_pts = np.split(points, i)
            x_data = np.array([self.x_func(a.T[0]) for a in gr_pts
                               if len(a) > 0])
            y_data = np.array([self.y_func(a.T[1]) for a in gr_pts
                               if len(a) > 0])
            assert len(x_data) == len(y_data)

            points = np.array((x_data, y_data)).T
```